File: report/export_pdf.mjs
#!/usr/bin/env node
/**
 * Export report/report.md -> report.pdf.
 *
 * Pipeline: marked (GFM markdown -> HTML) + KaTeX (math) + highlight.js
 * (code highlighting), then headless Chromium prints the HTML to PDF.
 * A real browser engine renders tables, Unicode, images and code properly.
 *
 * Usage:
 *   node report/export_pdf.mjs [--out report.pdf] [--keep-html]
 *   make pdf
 */

import { execFileSync } from "node:child_process";
import { existsSync } from "node:fs";
import { mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { createRequire } from "node:module";

import { Marked } from "marked";
import { markedHighlight } from "marked-highlight";
import hljs from "highlight.js";
import katex from "katex";

const require = createRequire(import.meta.url);
const HERE = dirname(fileURLToPath(import.meta.url));
const DEFAULT_MD = join(HERE, "report.md");
const DEFAULT_PDF = join(HERE, "..", "report.pdf");

// ---------------------------------------------------------------------------
// Markdown -> HTML (marked + KaTeX + highlight.js)
// ---------------------------------------------------------------------------

// KaTeX renderer: inline $...$ and display $$...$$.
// Guard against stray dollar signs (currency, ranges) by requiring the
// content to contain at least one backslash command or a known math token.
const MATH_RE = /\$\$([\s\S]+?)\$\$|\$([^$\n]+?)\$/g;

function renderMath(text) {
    return text.replace(MATH_RE, (whole, display, inline) => {
        const src = (display ?? inline).trim();
        if (!/\\[a-zA-Z]+|_|\^/.test(src)) return whole; // not math, keep as-is
        try {
            return katex.renderToString(src, {
                displayMode: Boolean(display),
                throwOnError: false,
                strict: false,
            });
        } catch {
            return whole;
        }
    });
}

const marked = new Marked(
    markedHighlight({
        langPrefix: "hljs language-",
        highlight(code, lang) {
            if (lang && hljs.getLanguage(lang)) {
                try {
                    return hljs.highlight(code, { language: lang }).value;
                } catch {
                    /* fall through */
                }
            }
            return hljs.highlightAuto(code).value;
        },
    })
);

// Custom renderer: force <details> open so content prints.
const renderer = {
    html({ text }) {
        return text.replace(/<details>/g, "<details open>");
    },
};
marked.use({ renderer });

function mdToHtml(mdText) {
    const raw = marked.parse(mdText, { async: false });
    return renderMath(raw);
}

// ---------------------------------------------------------------------------
// CSS (print-oriented)
// ---------------------------------------------------------------------------

const CSS = `
@page { size: A4; margin: 14mm 13mm; }
html { -webkit-print-color-adjust: exact; print-color-adjust: exact; }
body { font-family: 'DejaVu Sans', sans-serif; font-size: 9pt;
       line-height: 1.4; color: #1a1a1a; }
h1 { font-size: 17pt; border-bottom: 2px solid #333; padding-bottom: 4px; margin: 8px 0 10px; }
h2 { font-size: 12pt; border-bottom: 1px solid #ccc; padding-bottom: 2px;
     margin-top: 16px; margin-bottom: 6px; }
h3 { font-size: 10pt; margin-top: 10px; margin-bottom: 4px; }
table { border-collapse: collapse; width: 100%; font-size: 7.5pt; margin: 6px 0;
        table-layout: fixed; }
th, td { border: 1px solid #bbb; padding: 3px 4px; text-align: left;
         vertical-align: top; overflow-wrap: anywhere; }
th { background: #f0f0f0; }
pre { background: #f6f8fa; border: 1px solid #ddd; border-radius: 4px;
      padding: 8px; font-size: 8.5pt; white-space: pre-wrap;
      word-wrap: break-word; }
code { font-family: 'DejaVu Sans Mono', monospace; background: #f0f0f0;
       padding: 0 2px; border-radius: 3px; }
pre code { background: none; padding: 0; }
img { max-width: 100%; }
blockquote { border-left: 3px solid #ccc; margin: 8px 0; padding-left: 12px;
             color: #555; }
details { margin: 8px 0; }
summary { font-weight: bold; cursor: pointer; }
hr { border: none; border-top: 1px solid #ccc; margin: 18px 0; }
ul, ol { margin: 6px 0; padding-left: 22px; }
li { margin: 2px 0; }
a { color: #0366d6; }
/* KaTeX */
.katex { font-size: 1.05em; }
`;

// ---------------------------------------------------------------------------
// Chromium -> PDF
// ---------------------------------------------------------------------------

function findChromium() {
    for (const name of ["chromium", "chromium-browser", "google-chrome", "google-chrome-stable"]) {
        try {
            const path = execFileSync("which", [name], { encoding: "utf8" }).trim();
            if (path) return path;
        } catch {
            /* try next */
        }
    }
    return null;
}

function main() {
    const args = process.argv.slice(2);
    let outPdf = DEFAULT_PDF;
    let keepHtml = false;
    for (let i = 0; i < args.length; i++) {
        if (args[i] === "--out") outPdf = args[++i];
        else if (args[i] === "--keep-html") keepHtml = true;
        else if (args[i] === "--help" || args[i] === "-h") {
            console.log("Usage: node export_pdf.mjs [--out report.pdf] [--keep-html]");
            process.exit(0);
        }
    }

    const mdText = readFileSync(DEFAULT_MD, "utf8");
    const body = mdToHtml(mdText);

    const katexCss = require.resolve("katex/dist/katex.min.css");
    const hljsCss = require.resolve("highlight.js/styles/github.css");
    const katexCssText = readFileSync(katexCss, "utf8");
    const hljsCssText = readFileSync(hljsCss, "utf8");

    const htmlDoc = `<!DOCTYPE html>
<html><head><meta charset="utf-8">
<style>${katexCssText}</style>
<style>${hljsCssText}</style>
<style>${CSS}</style>
</head><body>${body}</body></html>`;

    const tmpDir = mkdtempSync(join(tmpdir(), "report-pdf-"));
    const tmpHtml = join(tmpDir, "report.html");
    writeFileSync(tmpHtml, htmlDoc, "utf8");

    const chromium = findChromium();
    if (!chromium) {
        console.error("error: no chromium/chrome found; cannot print to PDF");
        process.exit(1);
    }

    const outAbs = resolve(outPdf);
    const url = "file://" + tmpHtml;
    const cmd = [chromium, "--headless=new", "--disable-gpu", "--no-sandbox",
        "--no-pdf-header-footer", `--print-to-pdf=${outAbs}`, url];
    try {
        execFileSync(cmd[0], cmd.slice(1), { stdio: "inherit" });
    } catch {
        // Older Chromium: retry without --no-pdf-header-footer
        const retry = cmd.filter((a) => a !== "--no-pdf-header-footer");
        try {
            execFileSync(retry[0], retry.slice(1), { stdio: "inherit" });
        } catch (e) {
            console.error("error: chromium failed:\n" + (e.stderr || e.message));
            process.exit(1);
        }
    }

    if (!keepHtml) rmSync(tmpDir, { recursive: true, force: true });
    console.log(`ok: ${outAbs}`);
}

main();

File: report/export_pdf.py
#!/usr/bin/env python3
"""Export report/report.md -> report/report.pdf via headless Chromium.

No third-party dependencies: the markdown -> HTML step is a small,
self-contained converter, and the HTML -> PDF step uses the system
Chromium in headless print mode (a real browser engine, so tables,
Unicode, images and code blocks render properly).

Supported markdown subset (covers report.md today):
  - ATX headings (# .. ######)
  - paragraphs, bold/italic, inline code, links, images
  - fenced code blocks (```lang ... ```)
  - GFM pipe tables (header + --- separator + rows)
  - flat unordered lists (- / *)
  - blockquotes (>)
  - horizontal rules (---)
  - raw <details>/<summary> blocks (forced open so content is printed)
  - HTML comments (dropped)

NOT supported (script warns, does not crash): LaTeX math, footnotes,
definition lists, nested lists, Mermaid, task lists. For those, install
pandoc + a LaTeX engine and use: pandoc report.md -o report.pdf.

Usage:
    python3 report/export_pdf.py [--out report/report.pdf] [--keep-html]
"""

import argparse
import html
import os
import re
import shutil
import subprocess
import sys
import tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
DEFAULT_MD = os.path.join(HERE, "report.md")
DEFAULT_PDF = os.path.join(HERE, "../report.pdf")

# ---------------------------------------------------------------------------
# Markdown -> HTML
# ---------------------------------------------------------------------------

INLINE_RE = re.compile(
    r"(!\[[^\]]*\]\([^)]*\)"   # image ![alt](url)
    r"|\[[^\]]*\]\([^)]*\)"    # link  [text](url)
    r"|`[^`]*`"                # code  `...`
    r"|\*\*[^*]+\*\*"          # bold  **...**
    r"|\*[^*]+\*)"             # italic *...*
)


def inline(text):
    """Apply inline formatting; escape everything else and preserve
    raw HTML break tags by moving them to a private sentinel before
    escaping, then restoring them in the final HTML output.
    """
    # Support HTML `<br>` and `<br/>` inside Markdown prose or table cells.
    text = re.sub(r"<br\s*/?>", "__BR_TAG__", text, flags=re.IGNORECASE)

    out = []
    for part in INLINE_RE.split(text):
        if part is None:
            continue
        if part.startswith("![") and part.endswith(")"):
            m = re.match(r"!\[([^\]]*)\]\(([^)]*)\)", part)
            out.append(f'<img src="{html.escape(m.group(2))}" alt="{html.escape(m.group(1))}">')
        elif part.startswith("[") and part.endswith(")"):
            m = re.match(r"\[([^\]]*)\]\(([^)]*)\)", part)
            out.append(f'<a href="{html.escape(m.group(2))}">{inline(m.group(1))}</a>')
        elif part.startswith("`") and part.endswith("`"):
            out.append(f"<code>{html.escape(part[1:-1])}</code>")
        elif part.startswith("**") and part.endswith("**"):
            out.append(f"<strong>{inline(part[2:-2])}</strong>")
        elif part.startswith("*") and part.endswith("*"):
            out.append(f"<em>{inline(part[1:-1])}</em>")
        else:
            out.append(html.escape(part))

    return "".join(out).replace("__BR_TAG__", "<br>")


def parse_table(lines):
    """lines: consecutive '|'-prefixed lines. Returns HTML table.

    Apply inline formatting cell-by-cell so raw HTML line breaks inside
    a table cell survive as real `<br>` tags in the generated HTML.
    """
    rows = [l.strip().strip("|") for l in lines]
    cells = [[c.strip() for c in r.split("|")] for r in rows]
    header, body = cells[0], cells[2:]  # row 1 is the --- separator

    header_cells = [inline(c) for c in header]
    thead = "<tr>" + "".join(f"<th>{c}</th>" for c in header_cells) + "</tr>"

    tbody = ""
    for row in body:
        body_cells = [inline(c) for c in row]
        tbody += "<tr>" + "".join(f"<td>{c}</td>" for c in body_cells) + "</tr>"

    return f"<table><thead>{thead}</thead><tbody>{tbody}</tbody></table>"


def md_to_html(md_text, warnings):
    # Drop HTML comments (may span multiple lines).
    md_text = re.sub(r"<!--.*?-->", "", md_text, flags=re.DOTALL)

    lines = md_text.splitlines()
    out = []
    i = 0
    n = len(lines)
    while i < n:
        line = lines[i]
        stripped = line.strip()

        # Raw HTML passthrough (details/summary) — force details open.
        if stripped == "<details>":
            out.append("<details open>")
            i += 1
            continue
        if stripped == "</details>" or stripped.startswith("<summary>"):
            out.append(stripped)
            i += 1
            continue

        # Fenced code block
        m = re.match(r"^```(\w*)\s*$", stripped)
        if m:
            lang = m.group(1)
            buf = []
            i += 1
            while i < n and not lines[i].strip().startswith("```"):
                buf.append(lines[i])
                i += 1
            i += 1  # skip closing fence
            code = html.escape("\n".join(buf))
            cls = f' class="language-{lang}"' if lang else ""
            out.append(f"<pre><code{cls}>{code}</code></pre>")
            continue

        # Heading
        m = re.match(r"^(#{1,6})\s+(.*)$", stripped)
        if m:
            level = len(m.group(1))
            out.append(f"<h{level}>{inline(m.group(2))}</h{level}>")
            i += 1
            continue

        # Table: current line starts with '|' and next line is a separator
        if stripped.startswith("|") and i + 1 < n and re.match(r"^\|[\s:|-]+\|?\s*$", lines[i + 1].strip()):
            buf = [stripped]
            i += 1
            while i < n and lines[i].strip().startswith("|"):
                buf.append(lines[i].strip())
                i += 1
            out.append(parse_table(buf))
            continue

        # Horizontal rule
        if re.match(r"^-{3,}\s*$", stripped):
            out.append("<hr>")
            i += 1
            continue

        # Unordered list (flat)
        if re.match(r"^[-*]\s+", stripped):
            buf = []
            while i < n and re.match(r"^[-*]\s+", lines[i].strip()):
                buf.append(inline(re.sub(r"^[-*]\s+", "", lines[i].strip())))
                i += 1
            out.append("<ul>" + "".join(f"<li>{b}</li>" for b in buf) + "</ul>")
            continue

        # Blockquote
        if stripped.startswith(">"):
            buf = []
            while i < n and lines[i].strip().startswith(">"):
                buf.append(inline(re.sub(r"^>\s?", "", lines[i].strip())))
                i += 1
            out.append("<blockquote>" + "<br>".join(buf) + "</blockquote>")
            continue

        # Paragraph: accumulate until blank line or a block-starting line
        buf = []
        while i < n and lines[i].strip():
            s = lines[i].strip()
            if re.match(r"^(#{1,6})\s", s) or re.match(r"^```", s) or \
               re.match(r"^[-*]\s+", s) or s.startswith("|") or \
               re.match(r"^-{3,}\s*$", s) or s.startswith(">"):
                break
            buf.append(s)
            i += 1
        if buf:
            out.append("<p>" + inline(" ".join(buf)) + "</p>")
            continue

        i += 1  # blank line

    return "\n".join(out)


CSS = """
@page { size: A4; margin: 14mm 13mm; }
body { font-family: 'DejaVu Sans', sans-serif; font-size: 9pt;
       line-height: 1.4; color: #1a1a1a; }
h1 { font-size: 17pt; border-bottom: 2px solid #333; padding-bottom: 4px; margin: 8px 0 10px; }
h2 { font-size: 12pt; border-bottom: 1px solid #ccc; padding-bottom: 2px;
     margin-top: 16px; margin-bottom: 6px; }
h3 { font-size: 10pt; margin-top: 10px; margin-bottom: 4px; }
table { border-collapse: collapse; width: 100%; font-size: 7.5pt; margin: 6px 0;
        table-layout: fixed; }
th, td { border: 1px solid #bbb; padding: 3px 4px; text-align: left;
         vertical-align: top; overflow-wrap: anywhere; }
th { background: #f0f0f0; }
pre { background: #f6f8fa; border: 1px solid #ddd; border-radius: 4px;
      padding: 8px; font-size: 8.5pt; white-space: pre-wrap;
      word-wrap: break-word; }
code { font-family: 'DejaVu Sans Mono', monospace; background: #f0f0f0;
       padding: 0 2px; border-radius: 3px; }
pre code { background: none; padding: 0; }
img { max-width: 100%; }
blockquote { border-left: 3px solid #ccc; margin: 8px 0; padding-left: 12px;
             color: #555; }
details { margin: 8px 0; }
summary { font-weight: bold; cursor: pointer; }
hr { border: none; border-top: 1px solid #ccc; margin: 18px 0; }
ul { margin: 6px 0; padding-left: 22px; }
a { color: #0366d6; }
"""


def find_chromium():
    for name in ("chromium", "chromium-browser", "google-chrome", "google-chrome-stable"):
        path = shutil.which(name)
        if path:
            return path
    return None


def main():
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--out", default=DEFAULT_PDF, help="output PDF path")
    ap.add_argument("--keep-html", action="store_true", help="keep the intermediate HTML")
    args = ap.parse_args()

    with open(DEFAULT_MD, "r", encoding="utf-8") as f:
        md_text = f.read()

    warnings = []
    body = md_to_html(md_text, warnings)
    html_doc = f"<!DOCTYPE html><html><head><meta charset='utf-8'>" \
               f"<style>{CSS}</style></head><body>{body}</body></html>"

    tmp_html = os.path.join(HERE, ".report.html")
    with open(tmp_html, "w", encoding="utf-8") as f:
        f.write(html_doc)

    chromium = find_chromium()
    if not chromium:
        print("error: no chromium/chrome found; cannot print to PDF", file=sys.stderr)
        sys.exit(1)

    out_pdf = os.path.abspath(args.out)
    url = "file://" + tmp_html
    cmd = [chromium, "--headless=new", "--disable-gpu", "--no-sandbox",
           "--no-pdf-header-footer", f"--print-to-pdf={out_pdf}", url]
    r = subprocess.run(cmd, capture_output=True, text=True)
    if r.returncode != 0:
        # Older Chromium: retry without --no-pdf-header-footer
        cmd.remove("--no-pdf-header-footer")
        r = subprocess.run(cmd, capture_output=True, text=True)
    if r.returncode != 0:
        print("error: chromium failed:\n" + r.stderr[-2000:], file=sys.stderr)
        sys.exit(1)

    if not args.keep_html:
        os.remove(tmp_html)

    if warnings:
        print("warnings:")
        for w in warnings:
            print("  -", w)
    print(f"ok: {out_pdf}")


if __name__ == "__main__":
    main()